helix-term/src/ui/statusline.rs

```diff
@@ -0,0 +1,335 @@
+            helix_view::editor::StatusLineElement::FileEncoding => Self::render_file_encoding,
+            helix_view::editor::StatusLineElement::FileType => Self::render_file_type,
+            helix_view::editor::StatusLineElement::Diagnostics => Self::render_diagnostics,
+            helix_view::editor::StatusLineElement::Selections => Self::render_selections,
+            helix_view::editor::StatusLineElement::Position => Self::render_position,
+        }
+    }
+
+    fn render_mode<F>(context: &mut RenderContext, write: F)
+    where
+        F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
+    {
+        let visible = context.focused;
+
+        write(
+            context,
+            format!(
+                " {} ",
+                if visible {
+                    match context.doc.mode() {
+                        Mode::Insert => "INS",
+                        Mode::Select => "SEL",
+                        Mode::Normal => "NOR",
+                    }
+                } else {
+                    // If not focused, explicitly leave an empty space instead of returning None.
+                    "   "
+                }
+            ),
+            None,
+        );
+    }
+
+    fn render_lsp_spinner<F>(context: &mut RenderContext, write: F)
+    where
+        F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
+    {
+        write(
+            context,
+            format!(
+                " {} ",
+                context
+                    .doc
+                    .language_server()
+                    .and_then(|srv| context
+                        .spinners
+                        .get(srv.id())
+                        .and_then(|spinner| spinner.frame()))
+                    // Even if there's no spinner; reserve its space to avoid elements frequently shifting.
+                    .unwrap_or(" ")
+            ),
+            None,
+        );
+    }
+
+    fn render_diagnostics<F>(context: &mut RenderContext, write: F)
+    where
+        F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
+    {
+        let (warnings, errors) =
+            context
+                .doc
+                .diagnostics()
+                .iter()
+                .fold((0, 0), |mut counts, diag| {
+                    use helix_core::diagnostic::Severity;
+                    match diag.severity {
+                        Some(Severity::Warning) => counts.0 += 1,
+                        Some(Severity::Error) | None => counts.1 += 1,
+                        _ => {}
+                    }
+                    counts
+                });
+
+        if warnings > 0 {
+            Self::render_diagnostics_warning_state(context, write);
+            Self::render_diagnostics_warning_count(context, warnings, write);
+        }
+
+        if errors > 0 {
+            Self::render_diagnostics_error_state(context, write);
+            Self::render_diagnostics_error_count(context, errors, write);
+        }
+    }
```

Comment:
The contents of these functions can be inlined, I don't see the need for splitting them like this?

Reply:
Changed with c0a1870.